socketio/src/asynchronous/client/client.rs

```diff
@@ -26,49 +35,121 @@ use crate::{
     /// namespace. If `None` is passed in as namespace, the default namespace
     /// `"/"` is taken.
     /// ```
-    pub(crate) fn new<T: Into<String>>(
-        socket: InnerSocket,
-        namespace: T,
-        on: HashMap<Event, Callback<DynAsyncCallback>>,
-        on_any: Option<Callback<DynAsyncAnyCallback>>,
-        auth: Option<serde_json::Value>,
-    ) -> Result<Self> {
+    pub(crate) fn new(socket: InnerSocket, builder: ClientBuilder) -> Result<Self> {
         Ok(Client {
-            socket,
-            nsp: namespace.into(),
-            on: Arc::new(RwLock::new(on)),
-            on_any: Arc::new(RwLock::new(on_any)),
+            socket: Arc::new(RwLock::new(socket)),
+            nsp: builder.namespace.to_owned(),
             outstanding_acks: Arc::new(RwLock::new(Vec::new())),
-            auth,
+            auth: builder.auth.clone(),
+            builder: Arc::new(RwLock::new(builder)),
+            manually_disconnected: Arc::new(AtomicBool::new(false)),
+            server_disconnected: Arc::new(AtomicBool::new(false)),
         })
     }
 
     /// Connects the client to a server. Afterwards the `emit_*` methods can be
     /// called to interact with the server.
     pub(crate) async fn connect(&self) -> Result<()> {
         // Connect the underlying socket
-        self.socket.connect().await?;
+        self.socket.read().await.connect().await?;
 
         // construct the opening packet
         let auth = self.auth.as_ref().map(|data| data.to_string());
         let open_packet = Packet::new(PacketId::Connect, self.nsp.clone(), auth, None, 0, None);
 
-        self.socket.send(open_packet).await?;
+        self.socket.read().await.send(open_packet).await?;
+
+        Ok(())
+    }
+
+    pub(crate) async fn reconnect(&mut self) -> Result<()> {
+        let builder = self.builder.write().await;
+        let socket = builder.inner_create().await?;
+
+        // New inner socket that can be connected
+        let mut client_socket = self.socket.write().await;
+        *client_socket = socket;
+        drop(client_socket);
+
+        self.connect().await?;
+
+        Ok(())
+    }
+
+    /// Drives the stream using a thread so messages are processed
+    pub(crate) async fn poll_stream(&mut self) -> Result<()> {
+        let builder = self.builder.read().await;
+        let reconnect_delay_min = builder.reconnect_delay_min;
+        let reconnect_delay_max = builder.reconnect_delay_max;
+        let max_reconnect_attempts = builder.max_reconnect_attempts;
+        drop(builder);
+
+        let mut client_clone = self.clone();
+
+        tokio::runtime::Handle::current().spawn(async move {
+            loop {
+                let mut stream = client_clone.as_stream();
+                // Consume the stream until it returns None and the stream is closed.
+                while let Some(item) = stream.next().await {
+                    if let e @ Err(Error::IncompleteResponseFromEngineIo(_)) = item {
```

Comment:
Actually it might make sense to just trace any error here:
```
if let Err(err) = item {
  trace!("Error while polling packet stream: {err}");
}
```

Reply:
Makes sense. Updated.